# Modified 3D PMRNN

## Module Imports

In [1]:
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
config = tf.compat.v1.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.compat.v1.Session(config=config)
import cv2
import numpy as np
import matplotlib.pyplot as plt
from CONFIGURATION import CONFIGURATION
from Reconstruction import Reconstruction

In [2]:
# ignorable
print("TensorFlow version:", tf.__version__)
print("cv2 version:",cv2.__version__)
print("np version:",np.__version__)

TensorFlow version: 2.7.0
cv2 version: 4.5.5
np version: 1.19.5


In [3]:
# ignorable
import sklearn, matplotlib
print("matplotlib version:",matplotlib.__version__)
print("sklearn version:",sklearn.__version__)

matplotlib version: 3.6.3
sklearn version: 1.2.2


In [4]:
from tensorflow.keras import layers,losses #,Sequential,metrics
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import * 
# from tensorflow.keras.layers import Layer
# from tensorflow.keras import optimizers 
# from tensorflow.keras.optimizers.experimental import SGD
# from tensorflow.image import ssim

## Load new Data

### reading data

In [ ]:
import Utils
image_data = Utils.read_data("Berea_2d25um_binary.raw")
image_data.shape

### Voxel extraction 

In [ ]:
voxels = Utils.extract_subvolumes(image_data)
voxels.shape

In [ ]:
del image_data

### Train and test data extraction

In [ ]:
X_train = voxels[:-6]
X_test = voxels[-6:]

In [ ]:
del voxels

In [ ]:
len(X_train)

### Save Traited data for uses

In [ ]:
with open('X_train.npy', 'wb') as f:
    np.save(f, X_train)

In [ ]:
with open('X_test.npy', 'wb') as f:
    np.save(f, X_test)

## TensorFlow GPU setup

In [5]:
from tensorflow.python.compiler.tensorrt import trt_convert as trt
device = tf.config.list_physical_devices("GPU")
print(device)  # Check if GPU devices are visible
print(tf.test.is_built_with_cuda())  # Check if TensorFlow is built with CUDA support
print(trt.trt_utils._pywrap_py_utils.get_linked_tensorrt_version())

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
True
(7, 2, 2)


In [6]:
# only on weak GPU
tf.config.experimental.set_memory_growth(device[0],True)
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'

In [6]:
tf.keras.backend.clear_session()

## Load Traited data for training (voxels)

In [30]:
X_train=np.load('X_train.npy')
X_train.shape

(58, 250, 256, 256, 1)

In [31]:
# for test only
X_train=X_train[:2]
with open('X_train_test.npy', 'wb') as f:
    np.save(f, X_train)
X_train.shape

(2, 250, 256, 256, 1)

In [7]:
X_train=np.load('X_train_test.npy')
X_train.shape

(2, 250, 256, 256, 1)

### idk

In [8]:
learnedVoxels = []
inferenceVoxels = []
for voxel in X_train:
    learnedVoxels.append(voxel[:-1])
    inferenceVoxels.append(voxel[1:])

# learnedVoxels = np.array(learnedVoxels)
# inferenceVoxels = np.array(inferenceVoxels)

## Training 

### Imports

In [9]:
from mealpy.swarm_based import GWO
from sklearn.preprocessing import LabelEncoder
import time
import pickle
from datetime import datetime

In [10]:
import mealpy
mealpy.__version__

'2.4.2'

### Lables

In [11]:
class Timer():
    def __init__(self,title=" ",p={}) -> None:
        self.title=title
        self.started=time.time()
        self.params=p
        print(f"{self.title} Started at : {datetime.fromtimestamp(self.started)}")
        self.ended=None
    def End(self):
        self.ended=time.time()
        print(f"{self.title} Ended at : {datetime.fromtimestamp(self.ended)} and it took {self.ended-self.started}s")
    def setParams(self,name,data):
        self.params[name]=data

    def toDuration(self):
        if self.ended == None:
            return -1
        return self.ended-self.started
    
    def __str__(self) -> str:
        return f"{self.title} Started at : {datetime.fromtimestamp(self.started)} | Ended at : {datetime.fromtimestamp(self.ended)} and it took {self.ended-self.started}s"

class TrainingData(Timer):
    def __init__(self,title=" ",p={}) -> None:
        super().__init__(title=title,p=p)
        self.samples=[]
    
    def SampleStarted(self,title=" ")->Timer:
        timer= Timer(title)
        self.samples.append(timer)
        return timer
    
    def __str__(self) -> str:
        return super().__str__()+" "+str(self.params)

In [12]:
OPT_ENCODER = LabelEncoder()
#OPT_ENCODER.fit(['SGD', 'RMSprop', 'Adagrad', 'Adadelta', 'Adam', 'Adamax', 'Nadam'])
OPT_ENCODER.fit(['SGD', 'RMSprop', 'Adagrad', 'Adam'])
trainings_duration=[]
trainings_epochs=[]
current_iteration = 0


### Solution Decodation function

In [13]:
def decode_solution(solution):
    batch_size = 2**int(solution[0])
    
    learning_rate = solution[1]
    
    opt_int = int(solution[2])
    opt = OPT_ENCODER.inverse_transform([opt_int])[0]
    
    epoch = 10 *int(solution[3])
    
    num_filters = int(solution[4])
    
    latent_space_dim = 2**int(solution[5])
    
    reduced_dim = 2**int(solution[6])
    
    return [batch_size,learning_rate,opt,epoch,num_filters,latent_space_dim,reduced_dim]

### Objective Function

In [14]:
""" import random
# the range of float values
start = 0.001
end = 0.8 """
current_training=0

def objective_function(solution):
    tf.keras.backend.clear_session()
    global current_training
    print("\n###########################################################\n")
    print(f"\n--------------------------{current_iteration}/{current_training}--------------------------------\n")
    print("\n###########################################################\n")
    batch_size, learning_rate, opt, epoch, num_filters, latent_space_dim, reduced_dim = decode_solution(solution)
    reconstruction = Reconstruction(inputShape=CONFIGURATION["INPUT_SHAPE"],
                                    latent_space_dim=latent_space_dim,
                                    reducedDimension=reduced_dim,
                                    num_conv_layers=num_filters,
                                    learning_rate=learning_rate,
                                    batch_size=batch_size,
                                    epochs=epoch,
                                    opt=opt,
                                    )
    # reconstruction.summary()
    reconstruction.compile()
    histories = []
    params = {
        "inputShape": CONFIGURATION["INPUT_SHAPE"],
        "latent_space_dim": latent_space_dim,
        "reducedDimension": reduced_dim,
        "num_conv_layers": num_filters,
        "learning_rate": learning_rate,
        "batch_size": batch_size,
        "epochs": epoch,
        "opt": opt,
    }
    trainingTime = TrainingData("Epoch Training", params)
    for i in range(len(learnedVoxels)):
        print(f"currently working one voxel : {i +1}, voxels left : {len(learnedVoxels)-(i+1)} ")
        sampleTime = trainingTime.SampleStarted(f"Voxel Training {i+1}")
        # histories.append(random.uniform(start, end))
        histories.append(reconstruction.train(learnedVoxels[i], inferenceVoxels[i]))
        sampleTime.End()
    trainingTime.setParams("loss",histories[-1].history['loss'][-1])
    trainingTime.End()
    trainings_duration.append(trainingTime)

    current_training+=1
    
    """ reconstruction.save( save_folder=f"results/model_{batch_size}_{learning_rate}_{opt}_{epoch}_{num_filters}_{latent_space_dim}_{reduced_dim}") """
    #return min(histories, key=lambda x: float('inf') if x is None else x)
    return histories[-1].history['loss'][-1]

### Configurations

In [15]:
LB = [3    , 0.001 , 0   , 2      ,  4  ,  6    , 7 ]
UB = [5.99 , 0.01  , 3.99, 4.99  ,  8.99,  8.99 , 9.99 ]

problem_dict = {
    "fit_func": objective_function,
    "lb": LB,
    "ub": UB,
    "minmax": "min",
    "verbose":True,
}
metaheuristic_configuration={
    "epoch":5,
    "pop_size":10
}
#save session for x epoch of the training
save_interval=1
session_file = 'gwo_session.pkl'


### Lunching the metaheuristic

In [16]:
# Check if a session file exists
if os.path.exists(session_file):
    # Load the session and resume from the saved state
    with open(session_file, 'rb') as f:
        session_data = pickle.load(f)
    model = session_data['model']
    best_solution = session_data['best_solution']
    current_iteration = session_data['current_iteration']
    print(best_solution)
    print("Resuming from iteration", current_iteration)
    # raise ValueError("i just wanted to stop")
else:
    # Create a new GWO optimization session
    model = GWO.BaseGWO(problem_dict, epoch=metaheuristic_configuration["epoch"], pop_size=metaheuristic_configuration["pop_size"])
    best_solution = None
    current_iteration = 0
    trainings_epochs.append({"epoch" :current_iteration,"trainings" :trainings_duration,"best_solution":best_solution})
    trainings_duration=[]

{'fit': 0.01349127304975886, 'position': array([3.92447043e+00, 4.53517989e-03, 2.24030586e+00, 4.35225047e+00,
       8.53341056e+00, 6.79257622e+00, 7.66029412e+00])}
Resuming from iteration 2


In [17]:
# Run the optimization loop
#while current_iteration < metaheuristic_configuration["epoch"]:
    # Perform GWO iteration
best_position, best_fitness = model.solve()

# Update the best solution
if best_solution is None or best_fitness < best_solution["fit"]:
    best_solution = {
        "fit": best_fitness,
        "position": best_position
    }

current_iteration += 1
trainings_epochs.append({"epoch" :current_iteration,"trainings" :trainings_duration,"best_solution":best_solution})
trainings_duration=[]

# Increment the iteration count
# Save the session periodically
with open(session_file, 'wb') as f:
    pickle.dump({
        "model": model,
        "best_solution": best_solution,
        "current_iteration": current_iteration
    }, f)


###########################################################


--------------------------2/0--------------------------------


###########################################################



Instructions for updating:
Colocations handled automatically by placer.


Epoch Training Started at : 2023-06-19 18:31:30.094643
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:31:30.094796
Train on 249 samples
Epoch 1/40


2023-06-19 18:31:47.305947: W tensorflow/core/common_runtime/bfc_allocator.cc:343] Garbage collection: deallocate free memory regions (i.e., allocations) so that we can re-allocate a larger region to avoid OOM due to memory fragmentation. If you see this message frequently, you are running near the threshold of the available device memory and re-allocation may incur great performance overhead. You may try smaller batch sizes to observe the performance impact. Set TF_ENABLE_GPU_GARBAGE_COLLECTION=false if you'd like to disable this feature.


249/249 [==============================] - 14s 57ms/sample - loss: 0.2041 - _calculate_reconstruction_loss: 0.0782 - _calculate_porosity_loss: 0.0749 - _calculate_kl_loss: 47.5681
Epoch 2/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0711 - _calculate_reconstruction_loss: 0.0292 - _calculate_porosity_loss: 0.0051 - _calculate_kl_loss: 36.7609
Epoch 3/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0641 - _calculate_reconstruction_loss: 0.0290 - _calculate_porosity_loss: 0.0037 - _calculate_kl_loss: 31.3892
Epoch 4/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0623 - _calculate_reconstruction_loss: 0.0291 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 29.6988
Epoch 5/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0623 - _calculate_reconstruction_loss: 0.0290 - _calculate_porosity_loss: 0.0033 - _calculate_kl_loss: 29.8923
Epoch 6/40
249/249 [==============================] - 2s 7ms/sam

Epoch Training Started at : 2023-06-19 18:34:11.111920
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:34:11.112037
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 10s 40ms/sample - loss: 0.1233 - _calculate_reconstruction_loss: 0.0566 - _calculate_porosity_loss: 0.0413 - _calculate_kl_loss: 23.3734
Epoch 2/40
249/249 [==============================] - 3s 13ms/sample - loss: 0.0450 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0054 - _calculate_kl_loss: 11.6747
Epoch 3/40
249/249 [==============================] - 3s 13ms/sample - loss: 0.0420 - _calculate_reconstruction_loss: 0.0275 - _calculate_porosity_loss: 0.0052 - _calculate_kl_loss: 9.4772
Epoch 4/40
249/249 [==============================] - 4s 14ms/sample - loss: 0.0408 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0054 - _calculate_kl_loss: 7.9516
Epoch 5/40
249/249 [==============================] - 3s 1

Epoch Training Started at : 2023-06-19 18:39:18.769541
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:39:18.769630
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 4s 16ms/sample - loss: 0.4312 - _calculate_reconstruction_loss: 0.1684 - _calculate_porosity_loss: 0.1963 - _calculate_kl_loss: 65.1234
Epoch 2/30
249/249 [==============================] - 1s 5ms/sample - loss: 0.3115 - _calculate_reconstruction_loss: 0.1032 - _calculate_porosity_loss: 0.1436 - _calculate_kl_loss: 63.7445
Epoch 3/30
249/249 [==============================] - 1s 5ms/sample - loss: 0.2556 - _calculate_reconstruction_loss: 0.0779 - _calculate_porosity_loss: 0.1138 - _calculate_kl_loss: 63.3366
Epoch 4/30
249/249 [==============================] - 1s 5ms/sample - loss: 0.2222 - _calculate_reconstruction_loss: 0.0650 - _calculate_porosity_loss: 0.0943 - _calculate_kl_loss: 62.7283
Epoch 5/30
249/249 [==============================] - 1s 5ms

Epoch Training Started at : 2023-06-19 18:40:54.196550
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:40:54.196654
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.5735 - _calculate_reconstruction_loss: 0.2946 - _calculate_porosity_loss: 0.2192 - _calculate_kl_loss: 57.4402
Epoch 2/40
249/249 [==============================] - 1s 6ms/sample - loss: 0.4421 - _calculate_reconstruction_loss: 0.1925 - _calculate_porosity_loss: 0.1918 - _calculate_kl_loss: 57.1742
Epoch 3/40
249/249 [==============================] - 1s 6ms/sample - loss: 0.3932 - _calculate_reconstruction_loss: 0.1573 - _calculate_porosity_loss: 0.1780 - _calculate_kl_loss: 57.1449
Epoch 4/40
249/249 [==============================] - 1s 6ms/sample - loss: 0.3557 - _calculate_reconstruction_loss: 0.1333 - _calculate_porosity_loss: 0.1653 - _calculate_kl_loss: 56.7620
Epoch 5/40
249/249 [==============================] - 1s 6ms

Epoch Training Started at : 2023-06-19 18:43:10.218262
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:43:10.218399


Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


Train on 249 samples
Epoch 1/20
249/249 [==============================] - 7s 28ms/sample - loss: 0.3891 - _calculate_reconstruction_loss: 0.1613 - _calculate_porosity_loss: 0.1911 - _calculate_kl_loss: 32.5328
Epoch 2/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.2547 - _calculate_reconstruction_loss: 0.0855 - _calculate_porosity_loss: 0.1351 - _calculate_kl_loss: 32.3767
Epoch 3/20
249/249 [==============================] - 4s 15ms/sample - loss: 0.2000 - _calculate_reconstruction_loss: 0.0642 - _calculate_porosity_loss: 0.1019 - _calculate_kl_loss: 31.8441
Epoch 4/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.1616 - _calculate_reconstruction_loss: 0.0523 - _calculate_porosity_loss: 0.0781 - _calculate_kl_loss: 30.6333
Epoch 5/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.1380 - _calculate_reconstruction_loss: 0.0456 - _calculate_porosity_loss: 0.0613 - _calculate_kl_loss: 30.4139
Epoch 6/20
249/249 [=========

Epoch Training Started at : 2023-06-19 18:45:43.491519
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:45:43.491646
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 10s 40ms/sample - loss: 0.1546 - _calculate_reconstruction_loss: 0.0593 - _calculate_porosity_loss: 0.0376 - _calculate_kl_loss: 55.8987
Epoch 2/40
249/249 [==============================] - 4s 16ms/sample - loss: 0.0853 - _calculate_reconstruction_loss: 0.0285 - _calculate_porosity_loss: 0.0055 - _calculate_kl_loss: 51.2056
Epoch 3/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0795 - _calculate_reconstruction_loss: 0.0285 - _calculate_porosity_loss: 0.0050 - _calculate_kl_loss: 45.7606
Epoch 4/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0695 - _calculate_reconstruction_loss: 0.0282 - _calculate_porosity_loss: 0.0048 - _calculate_kl_loss: 36.3643
Epoch 5/40
249/249 [==============================] - 4s

Epoch Training Started at : 2023-06-19 18:52:08.223308
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:52:08.223435
Train on 249 samples
Epoch 1/20
249/249 [==============================] - 7s 30ms/sample - loss: 0.1236 - _calculate_reconstruction_loss: 0.0501 - _calculate_porosity_loss: 0.0441 - _calculate_kl_loss: 27.6988
Epoch 2/20
249/249 [==============================] - 3s 13ms/sample - loss: 0.0545 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0048 - _calculate_kl_loss: 22.4087
Epoch 3/20
249/249 [==============================] - 3s 12ms/sample - loss: 0.0524 - _calculate_reconstruction_loss: 0.0272 - _calculate_porosity_loss: 0.0050 - _calculate_kl_loss: 20.0284
Epoch 4/20
249/249 [==============================] - 3s 13ms/sample - loss: 0.0479 - _calculate_reconstruction_loss: 0.0270 - _calculate_porosity_loss: 0.0048 - _calculate_kl_loss: 16.2374
Epoch 5/20
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 18:54:34.059870
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:54:34.059988
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 6s 22ms/sample - loss: 0.2102 - _calculate_reconstruction_loss: 0.0814 - _calculate_porosity_loss: 0.0721 - _calculate_kl_loss: 53.2465
Epoch 2/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.0743 - _calculate_reconstruction_loss: 0.0295 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 40.9608
Epoch 3/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.0728 - _calculate_reconstruction_loss: 0.0289 - _calculate_porosity_loss: 0.0034 - _calculate_kl_loss: 40.2002
Epoch 4/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.0690 - _calculate_reconstruction_loss: 0.0285 - _calculate_porosity_loss: 0.0030 - _calculate_kl_loss: 37.4745
Epoch 5/40
249/249 [==============================] - 2s 9ms

Epoch Training Started at : 2023-06-19 18:58:03.787586
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 18:58:03.787707
Train on 249 samples
Epoch 1/30
248/249 [============================>.] - ETA: 0s - loss: 0.1198 - _calculate_reconstruction_loss: 0.0478 - _calculate_porosity_loss: 0.0433 - _calculate_kl_loss: 28.7409

2023-06-19 18:58:21.803026: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.15GiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.
2023-06-19 18:58:21.803332: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 2.15GiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 11s 43ms/sample - loss: 0.1195 - _calculate_reconstruction_loss: 0.0469 - _calculate_porosity_loss: 0.0422 - _calculate_kl_loss: 27.8427
Epoch 2/30
249/249 [==============================] - 6s 23ms/sample - loss: 0.0567 - _calculate_reconstruction_loss: 0.0285 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 23.7451
Epoch 3/30
249/249 [==============================] - 6s 23ms/sample - loss: 0.0589 - _calculate_reconstruction_loss: 0.0280 - _calculate_porosity_loss: 0.0044 - _calculate_kl_loss: 25.8378
Epoch 4/30
249/249 [==============================] - 6s 22ms/sample - loss: 0.0580 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0041 - _calculate_kl_loss: 26.0142
Epoch 5/30
249/249 [==============================] - 6s 22ms/sample - loss: 0.0563 - _calculate_reconstruction_loss: 0.0267 - _calculate_porosity_loss: 0.0040 - _calculate_kl_loss: 25.0805
Epoch 6/30
249/249 [==============================] - 6s 22m

Epoch Training Started at : 2023-06-19 19:03:35.784083
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:03:35.784191
Train on 249 samples
Epoch 1/20
249/249 [==============================] - 7s 27ms/sample - loss: 0.1661 - _calculate_reconstruction_loss: 0.0570 - _calculate_porosity_loss: 0.0440 - _calculate_kl_loss: 63.1050
Epoch 2/20
249/249 [==============================] - 3s 11ms/sample - loss: 0.0805 - _calculate_reconstruction_loss: 0.0280 - _calculate_porosity_loss: 0.0055 - _calculate_kl_loss: 46.8453
Epoch 3/20
249/249 [==============================] - 3s 11ms/sample - loss: 0.0692 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 37.1415
Epoch 4/20
249/249 [==============================] - 3s 12ms/sample - loss: 0.0630 - _calculate_reconstruction_loss: 0.0271 - _calculate_porosity_loss: 0.0047 - _calculate_kl_loss: 31.3504
Epoch 5/20
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 19:05:53.598461
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:05:53.598574
Train on 249 samples
Epoch 1/20
249/249 [==============================] - 6s 26ms/sample - loss: 0.4713 - _calculate_reconstruction_loss: 0.2207 - _calculate_porosity_loss: 0.2207 - _calculate_kl_loss: 27.2931
Epoch 2/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.3868 - _calculate_reconstruction_loss: 0.1615 - _calculate_porosity_loss: 0.1965 - _calculate_kl_loss: 27.1744
Epoch 3/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.3418 - _calculate_reconstruction_loss: 0.1338 - _calculate_porosity_loss: 0.1787 - _calculate_kl_loss: 27.5341
Epoch 4/20
249/249 [==============================] - 4s 14ms/sample - loss: 0.3112 - _calculate_reconstruction_loss: 0.1175 - _calculate_porosity_loss: 0.1651 - _calculate_kl_loss: 26.8167
Epoch 5/20
249/249 [==============================] - 4s 

Epoch Training Started at : 2023-06-19 19:08:24.168626
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:08:24.168720
Train on 249 samples
Epoch 1/20
249/249 [==============================] - 6s 26ms/sample - loss: 0.1784 - _calculate_reconstruction_loss: 0.0820 - _calculate_porosity_loss: 0.0682 - _calculate_kl_loss: 24.2493
Epoch 2/20
249/249 [==============================] - 4s 15ms/sample - loss: 0.0517 - _calculate_reconstruction_loss: 0.0272 - _calculate_porosity_loss: 0.0052 - _calculate_kl_loss: 18.6170
Epoch 3/20
249/249 [==============================] - 4s 15ms/sample - loss: 0.0490 - _calculate_reconstruction_loss: 0.0266 - _calculate_porosity_loss: 0.0047 - _calculate_kl_loss: 17.2073
Epoch 4/20
249/249 [==============================] - 4s 15ms/sample - loss: 0.0485 - _calculate_reconstruction_loss: 0.0261 - _calculate_porosity_loss: 0.0042 - _calculate_kl_loss: 17.7274
Epoch 5/20
249/249 [==============================] - 4s 

Epoch Training Started at : 2023-06-19 19:11:12.816086
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:11:12.816195
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 10s 39ms/sample - loss: 0.3004 - _calculate_reconstruction_loss: 0.1090 - _calculate_porosity_loss: 0.1602 - _calculate_kl_loss: 26.6786
Epoch 2/30
249/249 [==============================] - 5s 18ms/sample - loss: 0.1644 - _calculate_reconstruction_loss: 0.0521 - _calculate_porosity_loss: 0.0841 - _calculate_kl_loss: 26.4379
Epoch 3/30
249/249 [==============================] - 4s 17ms/sample - loss: 0.1132 - _calculate_reconstruction_loss: 0.0380 - _calculate_porosity_loss: 0.0489 - _calculate_kl_loss: 25.7633
Epoch 4/30
249/249 [==============================] - 4s 18ms/sample - loss: 0.0896 - _calculate_reconstruction_loss: 0.0329 - _calculate_porosity_loss: 0.0301 - _calculate_kl_loss: 26.1075
Epoch 5/30
249/249 [==============================] - 4s

Epoch Training Started at : 2023-06-19 19:15:59.231039
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:15:59.231164
Train on 249 samples
Epoch 1/20
249/249 [==============================] - 6s 23ms/sample - loss: 0.2565 - _calculate_reconstruction_loss: 0.0882 - _calculate_porosity_loss: 0.1395 - _calculate_kl_loss: 26.6426
Epoch 2/20
249/249 [==============================] - 3s 13ms/sample - loss: 0.1410 - _calculate_reconstruction_loss: 0.0452 - _calculate_porosity_loss: 0.0698 - _calculate_kl_loss: 24.9422
Epoch 3/20
249/249 [==============================] - 3s 13ms/sample - loss: 0.1072 - _calculate_reconstruction_loss: 0.0366 - _calculate_porosity_loss: 0.0444 - _calculate_kl_loss: 24.9226
Epoch 4/20
249/249 [==============================] - 3s 13ms/sample - loss: 0.0910 - _calculate_reconstruction_loss: 0.0330 - _calculate_porosity_loss: 0.0305 - _calculate_kl_loss: 26.0227
Epoch 5/20
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 19:18:23.403380
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:18:23.403471
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 26ms/sample - loss: 0.4418 - _calculate_reconstruction_loss: 0.1341 - _calculate_porosity_loss: 0.1954 - _calculate_kl_loss: 111.1793
Epoch 2/40
249/249 [==============================] - 2s 8ms/sample - loss: 0.3554 - _calculate_reconstruction_loss: 0.0913 - _calculate_porosity_loss: 0.1506 - _calculate_kl_loss: 113.2233
Epoch 3/40
249/249 [==============================] - 2s 8ms/sample - loss: 0.3014 - _calculate_reconstruction_loss: 0.0704 - _calculate_porosity_loss: 0.1197 - _calculate_kl_loss: 110.8884
Epoch 4/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.2667 - _calculate_reconstruction_loss: 0.0583 - _calculate_porosity_loss: 0.0975 - _calculate_kl_loss: 110.5350
Epoch 5/40
249/249 [==============================] - 2s

Epoch Training Started at : 2023-06-19 19:21:47.026275
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:21:47.026415
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 8s 32ms/sample - loss: 0.3763 - _calculate_reconstruction_loss: 0.1223 - _calculate_porosity_loss: 0.1874 - _calculate_kl_loss: 63.9813
Epoch 2/40
249/249 [==============================] - 4s 15ms/sample - loss: 0.2605 - _calculate_reconstruction_loss: 0.0694 - _calculate_porosity_loss: 0.1273 - _calculate_kl_loss: 63.3674
Epoch 3/40
249/249 [==============================] - 4s 15ms/sample - loss: 0.2058 - _calculate_reconstruction_loss: 0.0509 - _calculate_porosity_loss: 0.0916 - _calculate_kl_loss: 62.9659
Epoch 4/40
249/249 [==============================] - 4s 15ms/sample - loss: 0.1730 - _calculate_reconstruction_loss: 0.0420 - _calculate_porosity_loss: 0.0689 - _calculate_kl_loss: 62.0680
Epoch 5/40
249/249 [==============================] - 4s 

Epoch Training Started at : 2023-06-19 19:27:20.578663
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:27:20.578800
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 30ms/sample - loss: 0.2241 - _calculate_reconstruction_loss: 0.1057 - _calculate_porosity_loss: 0.0576 - _calculate_kl_loss: 56.9965
Epoch 2/40
249/249 [==============================] - 3s 10ms/sample - loss: 0.0825 - _calculate_reconstruction_loss: 0.0279 - _calculate_porosity_loss: 0.0053 - _calculate_kl_loss: 49.1925
Epoch 3/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0695 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 38.0894
Epoch 4/40
249/249 [==============================] - 3s 10ms/sample - loss: 0.0671 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0043 - _calculate_kl_loss: 35.2912
Epoch 5/40
249/249 [==============================] - 2s 

Epoch Training Started at : 2023-06-19 19:31:20.136328
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:31:20.136427
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 6s 25ms/sample - loss: 0.2990 - _calculate_reconstruction_loss: 0.1104 - _calculate_porosity_loss: 0.0836 - _calculate_kl_loss: 100.9478
Epoch 2/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.1313 - _calculate_reconstruction_loss: 0.0331 - _calculate_porosity_loss: 0.0068 - _calculate_kl_loss: 91.0825
Epoch 3/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.1187 - _calculate_reconstruction_loss: 0.0299 - _calculate_porosity_loss: 0.0051 - _calculate_kl_loss: 83.5648
Epoch 4/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.1069 - _calculate_reconstruction_loss: 0.0296 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 73.3802
Epoch 5/40
249/249 [==============================] - 2s 9m

Epoch Training Started at : 2023-06-19 19:34:50.532670
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:34:50.532785
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 6s 24ms/sample - loss: 0.1017 - _calculate_reconstruction_loss: 0.0481 - _calculate_porosity_loss: 0.0310 - _calculate_kl_loss: 20.7423
Epoch 2/30
249/249 [==============================] - 4s 15ms/sample - loss: 0.0485 - _calculate_reconstruction_loss: 0.0288 - _calculate_porosity_loss: 0.0036 - _calculate_kl_loss: 15.8155
Epoch 3/30
249/249 [==============================] - 4s 14ms/sample - loss: 0.0455 - _calculate_reconstruction_loss: 0.0280 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 13.3220
Epoch 4/30
249/249 [==============================] - 4s 14ms/sample - loss: 0.0450 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0050 - _calculate_kl_loss: 12.6256
Epoch 5/30
249/249 [==============================] - 4s 

Epoch Training Started at : 2023-06-19 19:39:04.001673
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:39:04.001804
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 10s 40ms/sample - loss: 0.6791 - _calculate_reconstruction_loss: 0.3220 - _calculate_porosity_loss: 0.2411 - _calculate_kl_loss: 112.7002
Epoch 2/40
249/249 [==============================] - 3s 13ms/sample - loss: 0.4933 - _calculate_reconstruction_loss: 0.1858 - _calculate_porosity_loss: 0.1937 - _calculate_kl_loss: 112.7260
Epoch 3/40
249/249 [==============================] - 3s 14ms/sample - loss: 0.4206 - _calculate_reconstruction_loss: 0.1376 - _calculate_porosity_loss: 0.1705 - _calculate_kl_loss: 111.9945
Epoch 4/40
249/249 [==============================] - 3s 12ms/sample - loss: 0.3712 - _calculate_reconstruction_loss: 0.1069 - _calculate_porosity_loss: 0.1517 - _calculate_kl_loss: 112.0469
Epoch 5/40
249/249 [==============================] 

Epoch Training Started at : 2023-06-19 19:43:00.242859
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:43:00.243041
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 10s 40ms/sample - loss: 0.1216 - _calculate_reconstruction_loss: 0.0397 - _calculate_porosity_loss: 0.0262 - _calculate_kl_loss: 54.5756
Epoch 2/40
249/249 [==============================] - 4s 18ms/sample - loss: 0.0761 - _calculate_reconstruction_loss: 0.0283 - _calculate_porosity_loss: 0.0054 - _calculate_kl_loss: 42.2418
Epoch 3/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0646 - _calculate_reconstruction_loss: 0.0281 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 31.6635
Epoch 4/40
249/249 [==============================] - 4s 18ms/sample - loss: 0.0616 - _calculate_reconstruction_loss: 0.0278 - _calculate_porosity_loss: 0.0046 - _calculate_kl_loss: 29.4314
Epoch 5/40
249/249 [==============================] - 4s

Epoch Training Started at : 2023-06-19 19:49:30.894097
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:49:30.894199
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 9s 37ms/sample - loss: 0.1200 - _calculate_reconstruction_loss: 0.0556 - _calculate_porosity_loss: 0.0346 - _calculate_kl_loss: 28.0999
Epoch 2/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0545 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0052 - _calculate_kl_loss: 21.7888
Epoch 3/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0516 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0049 - _calculate_kl_loss: 19.0870
Epoch 4/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0397 - _calculate_reconstruction_loss: 0.0273 - _calculate_porosity_loss: 0.0050 - _calculate_kl_loss: 7.7641
Epoch 5/40
249/249 [==============================] - 4s 1

Epoch Training Started at : 2023-06-19 19:55:49.010655
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 19:55:49.010770
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 11s 45ms/sample - loss: 0.1016 - _calculate_reconstruction_loss: 0.0450 - _calculate_porosity_loss: 0.0277 - _calculate_kl_loss: 26.9718
Epoch 2/30
249/249 [==============================] - 7s 30ms/sample - loss: 0.0578 - _calculate_reconstruction_loss: 0.0284 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 24.8885
Epoch 3/30
249/249 [==============================] - 6s 25ms/sample - loss: 0.0574 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0043 - _calculate_kl_loss: 25.0200
Epoch 4/30
249/249 [==============================] - 6s 23ms/sample - loss: 0.0576 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 25.0715
Epoch 5/30
249/249 [==============================] - 6s

Epoch Training Started at : 2023-06-19 20:01:30.061725
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:01:30.061832
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 7s 30ms/sample - loss: 0.1094 - _calculate_reconstruction_loss: 0.0473 - _calculate_porosity_loss: 0.0348 - _calculate_kl_loss: 25.9061
Epoch 2/30
249/249 [==============================] - 3s 13ms/sample - loss: 0.0538 - _calculate_reconstruction_loss: 0.0277 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 21.6120
Epoch 3/30
249/249 [==============================] - 3s 12ms/sample - loss: 0.0469 - _calculate_reconstruction_loss: 0.0275 - _calculate_porosity_loss: 0.0047 - _calculate_kl_loss: 14.4602
Epoch 4/30
249/249 [==============================] - 3s 13ms/sample - loss: 0.0457 - _calculate_reconstruction_loss: 0.0277 - _calculate_porosity_loss: 0.0053 - _calculate_kl_loss: 12.5335
Epoch 5/30
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 20:05:15.896566
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:05:15.896670
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 26ms/sample - loss: 0.1342 - _calculate_reconstruction_loss: 0.0572 - _calculate_porosity_loss: 0.0442 - _calculate_kl_loss: 30.8095
Epoch 2/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0588 - _calculate_reconstruction_loss: 0.0283 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 26.6277
Epoch 3/40
249/249 [==============================] - 3s 12ms/sample - loss: 0.0575 - _calculate_reconstruction_loss: 0.0280 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 26.0269
Epoch 4/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0570 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 26.0315
Epoch 5/40
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 20:09:34.563800
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:09:34.563922
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 30ms/sample - loss: 0.2484 - _calculate_reconstruction_loss: 0.0981 - _calculate_porosity_loss: 0.0601 - _calculate_kl_loss: 85.8208
Epoch 2/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0813 - _calculate_reconstruction_loss: 0.0283 - _calculate_porosity_loss: 0.0052 - _calculate_kl_loss: 47.5645
Epoch 3/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0546 - _calculate_reconstruction_loss: 0.0283 - _calculate_porosity_loss: 0.0048 - _calculate_kl_loss: 21.4594
Epoch 4/40
249/249 [==============================] - 3s 10ms/sample - loss: 0.0454 - _calculate_reconstruction_loss: 0.0282 - _calculate_porosity_loss: 0.0049 - _calculate_kl_loss: 12.3431
Epoch 5/40
249/249 [==============================] - 2s 

2023-06-19 20:11:11.159047: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0274 - _calculate_reconstruction_loss: 0.0219 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 2.0561
Epoch 32/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0259 - _calculate_reconstruction_loss: 0.0213 - _calculate_porosity_loss: 0.0031 - _calculate_kl_loss: 1.5423
Epoch 33/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0256 - _calculate_reconstruction_loss: 0.0208 - _calculate_porosity_loss: 0.0036 - _calculate_kl_loss: 1.0487
Epoch 34/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0258 - _calculate_reconstruction_loss: 0.0205 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 1.4646
Epoch 35/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0243 - _calculate_reconstruction_loss: 0.0197 - _calculate_porosity_loss: 0.0037 - _calculate_kl_loss: 0.8775
Epoch 36/40
249/249 [==============================] - 3s 11ms

2023-06-19 20:11:42.242673: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


192/249 [======================>.......] - ETA: 0s - loss: 0.0263 - _calculate_reconstruction_loss: 0.0214 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 0.9538

2023-06-19 20:11:43.277219: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0259 - _calculate_reconstruction_loss: 0.0213 - _calculate_porosity_loss: 0.0036 - _calculate_kl_loss: 1.0076
Epoch 4/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0237 - _calculate_reconstruction_loss: 0.0199 - _calculate_porosity_loss: 0.0029 - _calculate_kl_loss: 0.8773
Epoch 5/40
128/249 [==============>...............] - ETA: 1s - loss: 0.0234 - _calculate_reconstruction_loss: 0.0187 - _calculate_porosity_loss: 0.0036 - _calculate_kl_loss: 1.0590

2023-06-19 20:11:48.051629: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0233 - _calculate_reconstruction_loss: 0.0188 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 1.0275
Epoch 6/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0220 - _calculate_reconstruction_loss: 0.0179 - _calculate_porosity_loss: 0.0031 - _calculate_kl_loss: 0.9754
Epoch 7/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0218 - _calculate_reconstruction_loss: 0.0174 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 0.9308
Epoch 8/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0206 - _calculate_reconstruction_loss: 0.0165 - _calculate_porosity_loss: 0.0030 - _calculate_kl_loss: 1.0463
Epoch 9/40
 64/249 [======>.......................] - ETA: 1s - loss: 0.0204 - _calculate_reconstruction_loss: 0.0157 - _calculate_porosity_loss: 0.0036 - _calculate_kl_loss: 1.0259

2023-06-19 20:11:58.299419: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0204 - _calculate_reconstruction_loss: 0.0159 - _calculate_porosity_loss: 0.0039 - _calculate_kl_loss: 0.6671
Epoch 10/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0192 - _calculate_reconstruction_loss: 0.0152 - _calculate_porosity_loss: 0.0030 - _calculate_kl_loss: 1.0592
Epoch 11/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0207 - _calculate_reconstruction_loss: 0.0147 - _calculate_porosity_loss: 0.0032 - _calculate_kl_loss: 2.7434
Epoch 12/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0181 - _calculate_reconstruction_loss: 0.0145 - _calculate_porosity_loss: 0.0027 - _calculate_kl_loss: 0.9757
Epoch 13/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0184 - _calculate_reconstruction_loss: 0.0140 - _calculate_porosity_loss: 0.0031 - _calculate_kl_loss: 1.2524
Epoch 14/40
249/249 [==============================] - 3s 11ms

2023-06-19 20:12:24.698205: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0164 - _calculate_reconstruction_loss: 0.0124 - _calculate_porosity_loss: 0.0033 - _calculate_kl_loss: 0.6678
Epoch 19/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0153 - _calculate_reconstruction_loss: 0.0119 - _calculate_porosity_loss: 0.0023 - _calculate_kl_loss: 1.1232
Epoch 20/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0165 - _calculate_reconstruction_loss: 0.0119 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 1.1323
Epoch 21/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0152 - _calculate_reconstruction_loss: 0.0114 - _calculate_porosity_loss: 0.0024 - _calculate_kl_loss: 1.3885
Epoch 22/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0149 - _calculate_reconstruction_loss: 0.0111 - _calculate_porosity_loss: 0.0030 - _calculate_kl_loss: 0.8050
Epoch 23/40
249/249 [==============================] - 3s 11ms

2023-06-19 20:13:08.747735: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0133 - _calculate_reconstruction_loss: 0.0091 - _calculate_porosity_loss: 0.0030 - _calculate_kl_loss: 1.2050
Epoch 35/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0122 - _calculate_reconstruction_loss: 0.0088 - _calculate_porosity_loss: 0.0023 - _calculate_kl_loss: 1.1270
Epoch 36/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0122 - _calculate_reconstruction_loss: 0.0087 - _calculate_porosity_loss: 0.0027 - _calculate_kl_loss: 0.8543
Epoch 37/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0124 - _calculate_reconstruction_loss: 0.0086 - _calculate_porosity_loss: 0.0029 - _calculate_kl_loss: 0.9132
Epoch 38/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0118 - _calculate_reconstruction_loss: 0.0084 - _calculate_porosity_loss: 0.0026 - _calculate_kl_loss: 0.8701
Epoch 39/40
249/249 [==============================] - 3s 11ms

Epoch Training Started at : 2023-06-19 20:13:30.154619
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:13:30.154725
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 6s 23ms/sample - loss: 0.2025 - _calculate_reconstruction_loss: 0.0739 - _calculate_porosity_loss: 0.0757 - _calculate_kl_loss: 49.5850
Epoch 2/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0774 - _calculate_reconstruction_loss: 0.0313 - _calculate_porosity_loss: 0.0056 - _calculate_kl_loss: 40.1640
Epoch 3/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0695 - _calculate_reconstruction_loss: 0.0297 - _calculate_porosity_loss: 0.0056 - _calculate_kl_loss: 34.1635
Epoch 4/40
249/249 [==============================] - 2s 7ms/sample - loss: 0.0643 - _calculate_reconstruction_loss: 0.0292 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 31.4612
Epoch 5/40
249/249 [==============================] - 2s 7ms

Epoch Training Started at : 2023-06-19 20:16:19.818389
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:16:19.818499
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 9s 38ms/sample - loss: 0.1303 - _calculate_reconstruction_loss: 0.0475 - _calculate_porosity_loss: 0.0288 - _calculate_kl_loss: 52.5378
Epoch 2/40
249/249 [==============================] - 4s 17ms/sample - loss: 0.0736 - _calculate_reconstruction_loss: 0.0275 - _calculate_porosity_loss: 0.0047 - _calculate_kl_loss: 41.1820
Epoch 3/40
249/249 [==============================] - 4s 18ms/sample - loss: 0.0612 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0056 - _calculate_kl_loss: 27.9406
Epoch 4/40
249/249 [==============================] - 4s 18ms/sample - loss: 0.0605 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0056 - _calculate_kl_loss: 27.3973
Epoch 5/40
249/249 [==============================] - 4s 

Epoch Training Started at : 2023-06-19 20:22:49.733186
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:22:49.733310
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 30ms/sample - loss: 0.2206 - _calculate_reconstruction_loss: 0.0720 - _calculate_porosity_loss: 0.0620 - _calculate_kl_loss: 82.8525
Epoch 2/40
249/249 [==============================] - 3s 10ms/sample - loss: 0.0697 - _calculate_reconstruction_loss: 0.0281 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 36.3221
Epoch 3/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0505 - _calculate_reconstruction_loss: 0.0279 - _calculate_porosity_loss: 0.0047 - _calculate_kl_loss: 17.7411
Epoch 4/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0546 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0040 - _calculate_kl_loss: 24.4104
Epoch 5/40
249/249 [==============================] - 2s 

2023-06-19 20:26:28.878868: W tensorflow/core/common_runtime/bfc_allocator.cc:275] Allocator (GPU_0_bfc) ran out of memory trying to allocate 306.00MiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.


249/249 [==============================] - 3s 11ms/sample - loss: 0.0123 - _calculate_reconstruction_loss: 0.0090 - _calculate_porosity_loss: 0.0028 - _calculate_kl_loss: 0.6392
Epoch 35/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0113 - _calculate_reconstruction_loss: 0.0086 - _calculate_porosity_loss: 0.0023 - _calculate_kl_loss: 0.5185
Epoch 36/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0123 - _calculate_reconstruction_loss: 0.0088 - _calculate_porosity_loss: 0.0022 - _calculate_kl_loss: 1.3267
Epoch 37/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0124 - _calculate_reconstruction_loss: 0.0087 - _calculate_porosity_loss: 0.0028 - _calculate_kl_loss: 0.9643
Epoch 38/40
249/249 [==============================] - 3s 11ms/sample - loss: 0.0115 - _calculate_reconstruction_loss: 0.0085 - _calculate_porosity_loss: 0.0025 - _calculate_kl_loss: 0.5793
Epoch 39/40
249/249 [==============================] - 3s 11ms

Epoch Training Started at : 2023-06-19 20:26:50.898536
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:26:50.898647
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 6s 23ms/sample - loss: 0.5288 - _calculate_reconstruction_loss: 0.2002 - _calculate_porosity_loss: 0.2127 - _calculate_kl_loss: 113.4911
Epoch 2/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.3896 - _calculate_reconstruction_loss: 0.1135 - _calculate_porosity_loss: 0.1633 - _calculate_kl_loss: 111.9510
Epoch 3/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.3305 - _calculate_reconstruction_loss: 0.0870 - _calculate_porosity_loss: 0.1318 - _calculate_kl_loss: 111.1745
Epoch 4/40
249/249 [==============================] - 2s 9ms/sample - loss: 0.2913 - _calculate_reconstruction_loss: 0.0720 - _calculate_porosity_loss: 0.1085 - _calculate_kl_loss: 110.5932
Epoch 5/40
249/249 [==============================] - 2s

Epoch Training Started at : 2023-06-19 20:30:13.862688
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:30:13.862789
Train on 249 samples
Epoch 1/30
249/249 [==============================] - 7s 26ms/sample - loss: 0.1463 - _calculate_reconstruction_loss: 0.0652 - _calculate_porosity_loss: 0.0513 - _calculate_kl_loss: 27.3899
Epoch 2/30
249/249 [==============================] - 3s 11ms/sample - loss: 0.0535 - _calculate_reconstruction_loss: 0.0285 - _calculate_porosity_loss: 0.0038 - _calculate_kl_loss: 20.9792
Epoch 3/30
249/249 [==============================] - 3s 11ms/sample - loss: 0.0504 - _calculate_reconstruction_loss: 0.0282 - _calculate_porosity_loss: 0.0035 - _calculate_kl_loss: 18.7676
Epoch 4/30
249/249 [==============================] - 3s 12ms/sample - loss: 0.0504 - _calculate_reconstruction_loss: 0.0275 - _calculate_porosity_loss: 0.0037 - _calculate_kl_loss: 19.1550
Epoch 5/30
249/249 [==============================] - 3s 

Epoch Training Started at : 2023-06-19 20:33:31.202170
currently working one voxel : 1, voxels left : 1 
Voxel Training 1 Started at : 2023-06-19 20:33:31.202286
Train on 249 samples
Epoch 1/40
249/249 [==============================] - 7s 30ms/sample - loss: 0.2332 - _calculate_reconstruction_loss: 0.1015 - _calculate_porosity_loss: 0.0701 - _calculate_kl_loss: 57.6578
Epoch 2/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0815 - _calculate_reconstruction_loss: 0.0280 - _calculate_porosity_loss: 0.0056 - _calculate_kl_loss: 47.7071
Epoch 3/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0710 - _calculate_reconstruction_loss: 0.0274 - _calculate_porosity_loss: 0.0045 - _calculate_kl_loss: 38.7898
Epoch 4/40
249/249 [==============================] - 2s 10ms/sample - loss: 0.0639 - _calculate_reconstruction_loss: 0.0276 - _calculate_porosity_loss: 0.0059 - _calculate_kl_loss: 30.5398
Epoch 5/40
249/249 [==============================] - 2s 

2023-06-19 20:34:42.725319: W tensorflow/core/common_runtime/bfc_allocator.cc:462] Allocator (GPU_0_bfc) ran out of memory trying to allocate 128.00MiB (rounded to 134217728)requested by op training/RMSprop/gradients/gradients/AddN_2-0-TransposeNHWCToNCHW-LayoutOptimizer
If the cause is memory fragmentation maybe the environment variable 'TF_GPU_ALLOCATOR=cuda_malloc_async' will improve the situation. 
Current allocation summary follows.
Current allocation summary follows.
2023-06-19 20:34:42.726376: W tensorflow/core/common_runtime/bfc_allocator.cc:474] *********x**********************x*****************x*****x******x***x***********x******x*************
2023-06-19 20:34:42.726564: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at transpose_op.cc:183 : RESOURCE_EXHAUSTED: OOM when allocating tensor with shape[32,16,256,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc


ResourceExhaustedError: 2 root error(s) found.
  (0) RESOURCE_EXHAUSTED: OOM when allocating tensor with shape[32,16,256,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[{{node training/RMSprop/gradients/gradients/AddN_2-0-TransposeNHWCToNCHW-LayoutOptimizer}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info. This isn't available when running in Eager mode.

	 [[training/RMSprop/gradients/gradients/generate/batch_normalization_32/cond_grad/StatelessIf/else/_3527/gradients/FusedBatchNormV3_grad/FusedBatchNormGradV3/_3379]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info. This isn't available when running in Eager mode.

  (1) RESOURCE_EXHAUSTED: OOM when allocating tensor with shape[32,16,256,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[{{node training/RMSprop/gradients/gradients/AddN_2-0-TransposeNHWCToNCHW-LayoutOptimizer}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info. This isn't available when running in Eager mode.

0 successful operations.
0 derived errors ignored.

In [ ]:
# print("\n".join( [str(training) for epoch in trainings_epochs for training in epoch]))
print(f"Best solution : {best_solution['fit']}")
batch_size,learning_rate,opt,epoch,num_filters,latent_space_dim ,reduced_dim = decode_solution(best_solution["position"])
print(f"Batch size : {batch_size}, learning_rate : {learning_rate}, opt : {opt}, epoch : {epoch}, num_filters : {num_filters}, latent_space_dim : {latent_space_dim}, reduced_dim : {reduced_dim}")

Best solution : 0.01349127304975886
Batch size : 8, learning_rate : 0.0045351798887080024, opt : RMSprop, epoch : 40, num_filters : 8, latent_space_dim : 64, reduced_dim : 128


Best solution : 0.01349127304975886

Batch size : 8, learning_rate : 0.0045351798887080024, opt : RMSprop, epoch : 40, num_filters : 8, latent_space_dim : 64, reduced_dim : 128

In [28]:
current_iteration

2

### Save History

In [ ]:
model.history.save_global_objectives_chart(filename="hello/goc")
model.history.save_local_objectives_chart(filename="hello/loc")

model.history.save_global_best_fitness_chart(filename="hello/gbfc")
model.history.save_local_best_fitness_chart(filename="hello/lbfc")

model.history.save_runtime_chart(filename="hello/rtc")

model.history.save_exploration_exploitation_chart(filename="hello/eec")

model.history.save_diversity_chart(filename="hello/dc")

model.history.save_trajectory_chart(list_agent_idx=[3, 5], selected_dimensions=[3], filename="hello/tc")

### Remove old sessions

In [19]:
if os.path.exists(session_file):
    os.remove(session_file)

In [20]:
print(len(trainings_duration))
trainings_epochs.append({"epoch" :current_iteration,"trainings" :trainings_duration,"best_solution":best_solution})
trainings_duration=[]

31


In [21]:
for training in trainings_epochs:
    print(len(training["trainings"]))

31


In [22]:
import json
jd=[]
for training in trainings_epochs:
    json_data=[]
    for obj in training["trainings"]:
        training_obj=obj.__dict__
        samples=[]
        for sample in training_obj["samples"]:
            samples.append(sample.__dict__)
        training_obj["samples"]=samples
        json_data.append(training_obj)
    jd.append(json_data)  

# Print the JSON data
print(jd)
# Convert the object to JSON
with open("output.json", "w") as file:
    file.write(json.dumps(jd))

[[{'title': 'Epoch Training', 'started': 1687195890.0946434, 'params': {'inputShape': (256, 256, 1), 'latent_space_dim': 128, 'reducedDimension': 256, 'num_conv_layers': 7, 'learning_rate': 0.008538745818222471, 'batch_size': 32, 'epochs': 40, 'opt': 'Adam', 'loss': 0.011006402102370099}, 'ended': 1687196046.4754567, 'samples': [{'title': 'Voxel Training 1', 'started': 1687195890.094796, 'params': {}, 'ended': 1687195976.5112216}, {'title': 'Voxel Training 2', 'started': 1687195976.5114121, 'params': {}, 'ended': 1687196046.4753673}]}, {'title': 'Epoch Training', 'started': 1687196051.11192, 'params': {'inputShape': (256, 256, 1), 'latent_space_dim': 64, 'reducedDimension': 512, 'num_conv_layers': 7, 'learning_rate': 0.004769994103020501, 'batch_size': 16, 'epochs': 40, 'opt': 'RMSprop', 'loss': 0.01001345542031359}, 'ended': 1687196355.5249662, 'samples': [{'title': 'Voxel Training 1', 'started': 1687196051.1120374, 'params': {}, 'ended': 1687196207.6989279}, {'title': 'Voxel Training

## Testing the model

### Load test data

In [ ]:
X_test=np.load('X_test.npy')
X_test.shape

In [ ]:
first  = X_test[0]
Topredict = first[1:]
test = first[:-1]

In [ ]:
n = 10

plt.figure(figsize=(20,4))
for i in range(n):
    # original
    ax = plt.subplot(2,n,i+1)
    plt.imshow(Topredict[i].astype("float32"))
    plt.title('original')
    plt.gray()
    ax.get_xaxis().set_visible(False)
    ax.get_yaxis().set_visible(False)

    # reconstructed
    ax = plt.subplot(2,n,i+1+n)
    plt.imshow(testing[i].reshape(256,256,1))
    plt.title('reconstructed')
    plt.gray()
    ax.get_xaxis().set_visible(False)
    ax.get_yaxis().set_visible(False)

In [ ]:
# reconstruction.Generate.save('reconstruction_10batch_10epochs.h5')

In [ ]:
file_path = "reconstructed.raw"
testing.insert(0, first[0].reshape(1,256,256,1))
print(len(testing))

In [ ]:
final = np.array(testing,np.float32).reshape(250,256,256)
final.shape

In [ ]:
file_path = "my_array.npy"
np.save("file", final)

In [ ]:
final.tofile(file_path)

In [ ]:
Volume = np.fromfile(file_path, dtype=np.uint8)
Volume = Volume.reshape(250,256,256)
Volume.shape